### PR TITLE
Fix language service crash on startup

### PR DESCRIPTION
**Pull request by wmltogether**
We found a crash on startup in language service.
It happens when user preferred system language is not  supported or translated, LanguageService.CurrentLanguage should not be null on startup.

### CODE REVIEW

File: MicaForEveryone/Services/LanguageService.cs [hunk 1]
```diff
@@ -17,12 +17,25 @@ internal class LanguageService : ILanguageService
         public LanguageService()
         {
             SupportedLanguages = GetSupportedLanguages().ToArray();
-
-            var currentLanguage = ResourceManager.Current.DefaultContext.Languages[0];
-            CurrentLanguage = SupportedLanguages.First(
-                language => currentLanguage.StartsWith(language.LanguageTag));
+            foreach (var lstLanguage in Windows.System.UserProfile.GlobalizationPreferences.Languages)
```
Comment (dongle-the-gadget) [inline MicaForEveryone/Services/LanguageService.cs:20]: Have you tested this API in unpackaged builds?

**Comment by minusium** [comment]
Thanks for your PR. I've been aware of these bugs and actually I wanted to fix them now. I'll merge it.

> LanguageService.CurrentLanguage should not be null

Actually It's not null, `First(language => currentLanguage.StartsWith(language.LanguageTag)` throws an exception when nothing matches. My solution was using `FirstOrDefault` then it returns null when nothing matched, and when it's null I use English as fallback language.

File: MicaForEveryone.Win32/Window.cs [hunk 2]
```diff
@@ -284,8 +284,11 @@ public virtual void Dispose()
             _isDisposing = true;
             _ = NativeMethods.DestroyWindow(Handle);
             Handle = IntPtr.Zero;
-            Class.Dispose();
-            _ = NativeMethods.DestroyIcon(Class.Icon);
+            if (Class != null)
```
Comment (dongle-the-gadget) [inline MicaForEveryone.Win32/Window.cs:287]: I believe we should just skip Disposing if we don't have a Window Class.

File: MicaForEveryone/Services/LanguageService.cs (hunk 1 above)
Comment (minusium) [reply MicaForEveryone/Services/LanguageService.cs:20]: According to https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api It's not in the list of methods that require package identity.

Comment (dongle-the-gadget) [reply MicaForEveryone/Services/LanguageService.cs:20]: Never tried, and also I believe we can just simplify this by using LINQ.

File: MicaForEveryone.Win32/Window.cs (hunk 2 above)
Comment (minusium) [reply MicaForEveryone.Win32/Window.cs:287]: `Dispose` should be called after `Activate` method got called. We can add a flag and set it true when `Activate` called then in `Dispose` we check if this flag is true or not. calling `DestroyWindow` and `DestroyIcon` when `Handle` and `Class.Icon` are null is safe.

File: MicaForEveryone/Services/LanguageService.cs (hunk 1 above)
Comment (dongle-the-gadget) [reply MicaForEveryone/Services/LanguageService.cs:20]: The conversion to `Windows.Globalization.Language` is unnecessary as `Windows.System.UserProfile.GlobalizationPreferences.Languages` already returns language tags.

Comment (minusium) [reply MicaForEveryone/Services/LanguageService.cs:20]: I use `Language.NativeName` to show language name in Settings UI.

Comment (minusium) [reply MicaForEveryone/Services/LanguageService.cs:20]: > Never tried, and also I believe we can just simplify this by using LINQ.

I agree, I'm going to do that.

**Comment by wmltogether** [comment]
`ResourceManager.Current.DefaultContext.Languages[0] `will return the top preferred language (On my windows 11 , it will return zh-CN, cause I'm not a English user. But the 'zh-CN' locale is not supported so It will cause null, then application will crash ),
Actually , I have threee languages defined in Windows region settings ("zh-Hans-CN, en-US, ja"),  Windows.System.UserProfile.GlobalizationPreferences.Languages return the right language list, although 'zh-Hans-CN' is not supported,  but en-US (the second one ) is what I really need.

**Comment by minusium** [comment]
I thought it will return first user preferred language that is present in app's resources, if it was like this then this would work, but earlier I got a report from another user that had same problem.

**Comment by dongle-the-gadget** [comment]
> `ResourceManager.Current.DefaultContext.Languages[0] `will return the top preferred language (On my windows 11 , it will return zh-CN, cause I'm not a English user. But the 'zh-CN' locale is not supported so It will cause null, then application will crash ), Actually , I have threee languages defined in Windows region settings ("zh-Hans-CN, en-US, ja"), Windows.System.UserProfile.GlobalizationPreferences.Languages return the right language list, although 'zh-Hans-CN' is not supported, but en-US (the second one ) is what I really need.

In my understanding this bug is about the program not accounting for the full range of language preferences, I cannot see how this relates to the value being `null`.

**Comment by dongle-the-gadget** [comment]
I believe the crashing actually comes from the fact that we use the `First` LINQ function, which will throw an exception if there are no matching entities. We could replace that `First` function with `FirstOrDefault` and add a null check after that function, but that could cause the app to use English, even if the user actually prefers Spanish (a supported language) for example.

**Comment by wmltogether** [comment]
snipaste20220219221707
Sounds like a LinQ error (First(), or FirstOrDefault() will return null by default if nothing matched)

**Comment by dongle-the-gadget** [comment]
> snipaste20220219221707 Sounds like a LinQ error (First(), or FirstOrDefault() will return null by default if nothing matched)

It is. (btw `FirstOrDefault` actually returns the default value if there are no matches. In most cases it is `null`, but it might not in other cases, such as `default(int) = 0`)

**Comment by minusium** [comment]
> but that could cause the app to use English, even if the user actually prefers Spanish (a supported language) for example.

It won't be null when user's preferred language is in `SupportedLanguages` so it won't use fallback language.

File: MicaForEveryone/Services/LanguageService.cs [hunk 3]
```diff
@@ -17,24 +18,16 @@ internal class LanguageService : ILanguageService
...
-                CurrentLanguage = SupportedLanguages[0];
-            }
-        }
 
+            var preferredLanguageTag = GlobalizationPreferences.Languages.FirstOrDefault(
+                l => SupportedLanguages.Any(
+                    sl => l.ToLower().StartsWith(sl.LanguageTag.ToLower())));
+
+            var preferredLanguage = SupportedLanguages.FirstOrDefault(
+                l => preferredLanguageTag.ToLower().StartsWith(l.LanguageTag.ToLower()));
```
Comment (minusium) [inline MicaForEveryone/Services/LanguageService.cs]: ok I ruined it here again

**Review by minusium — APPROVED** [review]
That's nice too.

**Pull request merged**